Simplest solution to provide proper ordering.

slazicoicr opened the pull request:
This tries to address the issue seen in #217, where more than one point having the same X-axis coordinate prevents true sorting.

avarsava committed:
sorted in the DF, doesn't seem to sort in graph

avarsava committed:
it's called completed_date

avarsava committed:
You can't just ctrl-h it

avarsava committed:
I got confused

avarsava committed:
remove debug

slazicoicr committed:
Simplest solution to provide proper ordering.

slazicoicr committed:
[GR-1209] Fixes to bcl2barcode index calling (#237)

* Fixes to bcl2barcode index calling

* Mixed lanes are treated as single lanes

* Remove stray testing statement

* More why comments

avarsava reviewed (commented):
so I like the effect this has, but I think the extra number, especially displayed right beside the y value in the tooltip, is going to confuse users. Is having 2 rows with the same SampleName terribly common? Could we isolate just those ones and append a letter or even '_Run[n]' or something like that?

slazicoicr commented:
> 
> 
> so I like the effect this has, but I think the extra number, especially displayed right beside the y value in the tooltip, is going to confuse users. Is having 2 rows with the same SampleName terribly common? Could we isolate just those ones and append a letter or even '_Run[n]' or something like that?

Duplicated SampleName are the rule rather than the exception (most libraries are run on MiSeq and then on NovaSeq). Appending run info is going to make it extra long. Maybe:

* SampleName (1) or SampleName (ID1) or SampleName (PlotlyID1)
* Have the X-axis label be two lines using `<br>`. The second line then can be PlotlyID: 1 or the Run Alias

avarsava committed:
[GR-1205] Home Button on Navbar (#239)

Add a Home button to the navbar in the reports.
Change '.container' to static rather than relative, to fix bug with the Home button jumping down the page. I do not know why this works, or if this will have any consequences. If the navbar is acting weird, this is probably the commit you want to look at

slazicoicr committed:
[GR-1215] Bump to ETL 0.37.0 (bcl2barcode fix) (#240)

avarsava committed:
prepare release 200713-1430

avarsava committed:
prepare for next development iteration

avarsava commented:
We will discuss data labeling in the meeting tomorrow, then I will take over this PR

avarsava committed:
[GR-1200] Don't show timestamp on front page (#242)

avarsava committed:
Move to another column

avarsava committed:
[GR-1200] Grouped front page (#243)

avarsava committed:
prepare release 200720-1507

avarsava committed:
prepare for next development iteration

avarsava committed:
Demo working extra datalabel col

avarsava committed:
Refactor to make x_fn optional, Mode mandatory

avarsava committed:
Fix sorting being split by colour/shape

avarsava committed:
Fix sample name desync

avarsava committed:
Only use hovertext template if there's hovertext to show

avarsava committed:
Fix for ML reports

avarsava committed:
Convert all reports except single lane rna due to subplots'

avarsava committed:
Oh my god the subplots work

avarsava committed:
Cleanup pdb

avarsava commented:
@slazicoicr I've reworked quite a bit of this feature, in order to try to intelligently differentiate the x-axis from what we display as the x-axis. Please let me know if it needs more documentation, or if it doesn't make sense for some conditions (eg do we want this toggleable? If the First Sort is 'Sample' should it automatically toggle back?)

edit: Oh, and because the source branch isn't rebased on develop, the diff is terrible, sorry about that

slazicoicr commented on application/dash_application/utility/plot_builder.py:
```diff
@@ -264,8 +267,8 @@ def is_empty_plot(trace_list) -> bool:
-def generate(title_text, sorted_data, x_fn, y_fn, axis_text, colourby, shapeby,
-             hovertext_cols, cutoff_lines: List[Tuple[str, float]]=[],
+def generate(title_text, sorted_data, y_fn, axis_text, colourby, shapeby,
+             hovertext_cols, mode, x_fn=None, cutoff_lines: List[Tuple[str, float]]=[],
```
A lot of code already depends on this change, but I would not change the signature of this function. Rather I would keep `x_fn` in it's place and allow it to be the `Mode` enum. This way we avoid having two parameters that compete with each other for what the x-axis should be. This would also allowed existing code from views to work. The code would be

```
if x_fn == Mode.IUS:
    # Same as your current IUS code
elif x_fn == Mode.MERGED:
    # Same as your current MERGED code
else:
    # Same as your current else code
````

avarsava replied:
Though Python allows it, I would rather not keep the signature the same. My reasoning is that aside from accepting two different data types on the same formal parameter being unorthodox, it does less to communicate to other developers that the results of setting the Mode are the default behaviours, and defining the x_fn is an override rather than an equally valid option. Were they on equal footing I would fear confusion over why setting x_fn to 'lambda d: d[PINERY_COL.SampleName]' isn't equivalent to setting it to Mode.IUS, and it would be less discoverable that setting it with a Mode is expected.
I have in this PR already adjusted every call to generate() to use this signature, no further refactoring is required.

slazicoicr commented on application/dash_application/utility/plot_builder.py:
```diff
@@ -827,15 +846,29 @@ def __init__(
+                self.display_x = lambda d: d[ml_col]
+        else:
+            self.display_x = lambda d: x_fn(d)
+
```
Would is make sense to move above code to `_generate_traces` (rather than `generate`), to avoid this code duplication?

avarsava replied:
It is this way because the 'categoryarray' for the x-axis needs access to the correctly set x_fn.

slazicoicr replied:
Ok, my main fear was that `generate` and here could become divergent. I think that `generate` won't be called anymore once we switch over to subplots, so this is fine.

slazicoicr commented on application/dash_application/views/single_lane_rna.py:
```diff
@@ -187,26 +188,24 @@ def get_rna_data():
 ]
 
-
+# TODO: Aww beans the subplots are gonna give me grief
```
As all views will be switched over to subplots soon, the new ordering approach needs to work with subplots.

avarsava replied:
that TODO was outdated, sorry for leaving it in :sweat_smile:

avarsava committed:
Cleanup todo

slazicoicr commented on application/dash_application/utility/plot_builder.py:
```diff
@@ -827,15 +846,29 @@ def __init__(
+            if self.mode == Mode.IUS:
+                self.x_fn = lambda d: d["SampleNameExtra"]
+                self.display_x = lambda d: d[PINERY_COL.SampleName]
+            elif mode == Mode.MERGED:
```
Should be `self.mode`

avarsava replied:
Fixed

slazicoicr commented on application/dash_application/utility/plot_builder.py:
```diff
@@ -827,15 +846,29 @@ def __init__(
+                self.x_fn = lambda d: d[ml_col]
+                self.display_x = lambda d: d[ml_col]
+        else:
+            self.display_x = lambda d: x_fn(d)
```
Should be `self.x_fn`

avarsava replied:
Fixed

slazicoicr commented on application/dash_application/utility/plot_builder.py:
```diff
@@ -559,16 +573,20 @@ def _define_graph(data, x_fn, y_fn, bar_positive, bar_negative, hovertext_cols,
-
+    
+    additional_str = ""
+    if isinstance(hovertext, pandas.Series) and not hovertext.empty:
```
I think any Python sequence will work here  (will silently throw away `hovertext` if user supplied it in Python `list` format), `if len(hovertext) > 0` should work.

avarsava replied:
Fixed

slazicoicr reviewed (commented):
I think this will work (nice finding the `hovertemplate` solution). I think the cutoff lines hovertext needs to be fixed, as it shows the `SampleNameExtra` column data.

avarsava committed:
fix self

avarsava committed:
fix hovertext test'

avarsava commented:
thanks Savo, could I get your opinions on
> if it doesn't make sense for some conditions (eg do we want this toggleable? If the First Sort is 'Sample' should it automatically toggle back?)

before I start rebasing onto the changes re: subplots? Let me know if you'd like to meet about it

slazicoicr commented:
For now, I would leave it as it is (all points always have their own x-position). Points sharing an x-position causes problems with the plots that contain error bars, as the error bars from two points perfectly overlap. I would not include a toggle feature yet. Can be implemented on request.

I still need to switch two call ready views onto subplots, but I am fine waiting for this to be merged into develop.

avarsava commented:
Hi Savo, was this closed by #258 ?

slazicoicr commented:
Yes.
